src/theory/builtin/theory_builtin_rewriter.cpp

```diff
@@ -234,13 +237,75 @@ Node TheoryBuiltinRewriter::getArrayRepresentationForLambdaRec(TNode n,
       curr_val = curr[1];
       next = curr[2];
     }
+    else if (ck == kind::OR || ck == kind::AND)
+    {
+      Trace("builtin-rewrite-debug2")
+          << "  process base : " << curr << std::endl;
+      // curr = Rewriter::rewrite(curr);
+      // Trace("builtin-rewrite-debug2")
+      //     << "  rewriten base : " << curr << std::endl;
+      // Complex Boolean return cases, in which
+      //  (1) lambda x. (= x v1) v ... becomes
+      //      lambda x. (ite (= x v1) true [...])
+      //
+      //  (2) lambda x. (not (= x v1)) ^ ... becomes
+      //      lambda x. (ite (= x v1) false [...])
+      //
+      // Note the negated cases of the lhs of the OR/AND operators above are
+      // handled by pushing the recursion to the then-branch, with the
+      // else-branch being the constant value. For example, the negated (1)
+      // would be
+      //  (1') lambda x. (not (= x v1)) v ... becomes
+      //       lambda x. (ite (= x v1) [...] true)
+      // thus requiring the rest of the disjunction to be further processed in
+      // the then-branch as the current value.
+      bool pol = curr[0].getKind() != kind::NOT;
+      bool inverted = (pol && ck == kind::AND) || (!pol && ck == kind::OR);
+      index_eq = pol ? curr[0] : curr[0][0];
+      Node processed, remainder;
+      processed = nm->mkConst(pol && !inverted);
```

Comment:
Still confused at this value, since it prefers a value of "false".  Are you sure this is correct?  I would have perhaps expected `pol != inverted` ?

Reply:
Yes, this is not correct. If we are in the inverted case then the polarity is flipped, otherwise it is as computed. I'll update and add more doc, good catch.